src/core/router/index.php

```diff
@@ -0,0 +1,142 @@
+            $segmentsB = explode('/', $b['route']);
+            $paramsA = $paramsLen($segmentsA);
+            $paramsB = $paramsLen($segmentsB);
+
+            if ($paramsA !== $paramsB) return $paramsA - $paramsB;
+            return count($segmentsA) - count($segmentsB);
+        });
+    }
+
+    public function any($route, $controller) {
+        $this->routes[] = ['route' => $route, 'method' => null, 'controller' => $controller];
+        $this->sortRoutes();
+    }
+
+    public function get($route, $controller) {
+        $this->routes[] = ['route' => $route, 'method' => 'GET', 'controller' => $controller];
+        $this->sortRoutes();
+    }
+
+    public function post($route, $controller) {
+        $this->routes[] = ['route' => $route, 'method' => 'POST', 'controller' => $controller];
+        $this->sortRoutes();
+    }
+
+    public function put($route, $controller) {
+        $this->routes[] = ['route' => $route, 'method' => 'PUT', 'controller' => $controller];
+        $this->sortRoutes();
+    }
+
+    public function patch($route, $controller) {
+        $this->routes[] = ['route' => $route, 'method' => 'PATCH', 'controller' => $controller];
+        $this->sortRoutes();
+    }
+
+    public function delete($route, $controller) {
+        $this->routes[] = ['route' => $route, 'method' => 'DELETE', 'controller' => $controller];
+        $this->sortRoutes();
+    }
+
+    private function findSegments($route, $segmentsRoute, $segmentsUrl) {
+        if (count($segmentsRoute) !== count($segmentsUrl)) return null;
+
+        foreach ($segmentsRoute as $i => $segment) {
+            if (strpos($segment, ':') === 0) {
+                $key = substr($segment, 1);
+                $route['params'][$key] = $segmentsUrl[$i];
+            } else {
+                if ($segment !== $segmentsUrl[$i]) return null;
+            }
+        }
+
+        return $route;
+    }
+
+    private function findRoute($req) {
+        $pathname = strtok($req['url'], '?');
+        $segmentsUrl = explode('/', $pathname);
+
+        foreach ($this->routes as $route) {
+            $newRoute = $route;
+            $newRoute['params'] = [];
+            $segmentsRoute = explode('/', $route['route']);
+            $segments = $this->findSegments($newRoute, $segmentsRoute, $segmentsUrl);
+            if ($segments) return $newRoute;
+        }
+
+        return null;
+    }
+
+    private function getAgent($req) {
+        return $req['headers']['User-Agent'] ?? null;
+    }
+
+    private function getCookies($req) {
+        if (!isset($req['headers']['Cookie'])) return null;
+
+        $cookies = [];
+        foreach (explode('; ', $req['headers']['Cookie']) as $pair) {
+            list($key, $value) = explode('=', $pair, 2);
+            $cookies[$key] = $value;
+        }
+
+        return $cookies;
+    }
+
+    private function getParams($req) {
+        $query = [];
+        parse_str(parse_url($req['url'], PHP_URL_QUERY), $query);
```

Comment:
You need to ensure that parse_str is only called when the query string is not null.

The error message appeared:
Deprecated: parse_str(): Passing null to parameter #1 ($string) of type string is deprecated on line 103

Reply:
done